src/realm/object-store/sync/app.cpp

```diff
@@ -214,21 +215,45 @@ App::Config::DeviceInfo::DeviceInfo(std::string a_platform_version, std::string
 // NO_THREAD_SAFETY_ANALYSIS because clang generates a false positive.
 // "Calling function configure requires negative capability '!app->m_route_mutex'"
 // But 'app' is an object just created in this static method so it is not possible to annotate this in the header.
+#if REALM_ENABLE_SYNC
+SharedApp App::get_app(CacheMode mode, const Config& config, const SyncClientConfig& sync_client_config,
+                       std::shared_ptr<BackingStore> store) NO_THREAD_SAFETY_ANALYSIS
```

Comment:
Great points, passing a store instance and also having caching was broken. I implemented the store factory design and cleaned up the initialization of the SyncManager and RealmBackingStore. 

I think we do need to keep the SyncConfig separate from the RealmBackingStore constructor though. This allows SDKs to bring their own BackingStore implementation and have it work either with or without sync enabled. It is a bit messy to have the two overloads of `App::get_app` with and without sync, but I think it might be less dangerous than exposing `App::set_sync_manager()`?